neurodsp/aperiodic/dfa.py

```diff
@@ -0,0 +1,145 @@
+
+        - 1 for regular DFA
+        - 2 or higher for generalized DFA
+    method : {'dfa', 'rs'}
+        Method to use to compute fluctuations:
+
+        - 'dfa' : detrended fluctuation
+        - 'rs' : rescaled range
+
+    Returns
+    -------
+    t_scales : 1d array
+        Time-scales over which fluctuation measures were computed.
+    fluctuations : 1d array
+        Average fluctuation at each scale.
+    exp : float
+        Slope of line in loglog when plotting t_scales against fluctuations.
+        This is the alpha value for DFA, or the Hurst exponent for rescaled range.
+
+    Notes
+    -----
+    These analysis compute fractal properties by analyzing fluctuations across windows.
+
+    Overall, these approaches involve dividing the time-series into non-overlapping
+    windows at log-spaced scales and computing a fluctuation measure across windows.
+    The relationship of this fluctuation measure across window sizes provides
+    information on the fractal properties of a signal.
+
+    Measures available are:
+    - DFA: detrended fluctuation analysis
+        - computes ordinary least squares fits across signal windows
+    - RS: rescaled range
+        - computes the range of signal windows, divided by the standard deviation
+
+    Empirically, DFA seems to work for a larger range of exponent values.
+    """
+
+    # Get log10 equi-spaced scales and translate that into window lengths
+    t_scales = np.logspace(np.log10(min_scale), np.log10(max_scale), n_scales)
+    win_lens = np.round(t_scales * fs).astype('int')
+
+    # Calculate culmulative sum of the signal
+    sig_walk = sp.cumsum(sig - np.mean(sig))
+
+    # Step through each scale and get RMS
+    fluctuations = np.zeros_like(t_scales)
+    for idx, win_len in enumerate(win_lens):
+        if method is 'dfa':
+            fluctuations[idx] = compute_detrended_fluctuation(sig_walk, win_len=win_len, deg=deg)
+        elif method is 'rs':
+            fluctuations[idx] = compute_rescaled_range(sig_walk, win_len=win_len)
+        else:
+            raise ValueError('Fluctuation method not understood.')
+
+    # Calculate the relationship between between fluctuations & time scales
+    exp = np.polyfit(np.log10(t_scales), np.log10(fluctuations), deg=1)[0]
+
+    return t_scales, fluctuations, exp
+
+
+def compute_rescaled_range(sig, win_len):
+    """Compute rescaled range of a given time-series at a given scale.
+
+    Parameters
+    ----------
+    sig : 1d array
+        Time series.
+    win_len : int
+        Window length for each rescaled range computation.
+
+    Returns
+    -------
+    rs : float
+        Average rescaled range over windows.
+    """
+
+    # Gather windows & vectorize, so we can call math functions in one go
+    n_win = int(np.floor(len(sig) / win_len))
+    sig_rect = np.reshape(sig[:n_win * win_len], (n_win, win_len)).T
+
+    # get back the sig by taking the derivative of sig_walk
```

Comment:
This is just a way to re-compute the original signal, right? Am I missing something more to it?

It seems like more work than is needed, as opposed to just passing in the original signal. I suggest we just pass in the original signal, compute cumulative sum in here (see comment above), and then we don't need to re-compute it. 

Also: in practice this re-computes the demeaned version of the original signal. Which doesn't matter for computing std, but just seems a little roundabout (harder to read / understand the code, I find). Again - mostly just checking I'm not missing something here.

Reply:
Okay - so I reworked it to pass in the original signal, demean and calculate cumsum, and then resplit the demeaned signal and calculate std from there - so it still uses demeaned signal to calculated std, but doesn't have to recompute the original signal from the cumsum anymore.

Reply:
yeah I think that looks right. I'm following the math from here, and it looks like things are done in the right order: https://en.wikipedia.org/wiki/Detrended_fluctuation_analysis